archive/reader.go

```diff
@@ -1,175 +1,328 @@
-// Package archive is used to parse a .zip file retrieved
-// by the API.
-//
-// Here is the content of an archive retried on the tablet as example:
-// 384327f5-133e-49c8-82ff-30aa19f3cfa4.content
-// 384327f5-133e-49c8-82ff-30aa19f3cfa4//0-metadata.json
-// 384327f5-133e-49c8-82ff-30aa19f3cfa4//0.rm
-// 384327f5-133e-49c8-82ff-30aa19f3cfa4.pagedata
-// 384327f5-133e-49c8-82ff-30aa19f3cfa4.thumbnails/0.jpg
-//
-// As the .zip file from remarkable is simply a normal .zip file
-// containing specific file formats, this package is only a wrapper
-// around the standard zip package that will follow the same
-// code architecture and that will help gathering one of
-// those specific files more easily.
 package archive
 
 import (
 	"archive/zip"
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"errors"
 	"fmt"
+	"io"
+	"io/ioutil"
 	"path/filepath"
-	"sort"
 	"strconv"
 	"strings"
+
+	"github.com/juruen/rmapi/encoding/rm"
 )
 
-// A Page represents a note page.
-type Page struct {
-	Data      *zip.File
-	Metadata  *zip.File
-	Thumbnail *zip.File
+// Read fills a File struct parsing a Remarkable archive file.
+func (z *Zip) Read(r io.Reader) error {
+	zr, err := zipReaderFromIOReader(r)
+	if err != nil {
+		return err
+	}
+
+	// reading content first because it contains the number of pages
+	if err := z.readContent(zr); err != nil {
+		return err
+	}
+
+	// instanciate the slice of pages
+	if z.Content.PageCount == 0 {
+		return errors.New("document does not have any pages")
+	}
+	z.Pages = make([]Page, z.Content.PageCount)
+
+	if err := z.readPagedata(zr); err != nil {
+		return err
+	}
+
+	if err := z.readPdf(zr); err != nil {
+		return err
+	}
+
+	if err := z.readEpub(zr); err != nil {
+		return err
+	}
+
+	if err := z.readData(zr); err != nil {
+		return err
+	}
+
+	if err := z.readThumbnails(zr); err != nil {
+		return err
+	}
+
+	if err := z.readMetadata(zr); err != nil {
+		return err
+	}
+
+	return nil
 }
 
-// Reader will parse specific files of the remarkable zip file.
-type Reader struct {
-	Content  *zip.File
-	Pagedata *zip.File
-	Pages    []Page
-	Pdf      *zip.File
-	Epub     *zip.File
-	UUID     string
+// readContent reads the .content file contained in an archive
+func (z *Zip) readContent(zr *zip.Reader) error {
+	files, err := zipExtFinder(zr, ".content")
+	if err != nil {
+		return err
+	}
+
+	if len(files) != 1 {
+		return errors.New("archive does not contain a unique content file")
+	}
+
+	file, err := files[0].Open()
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	bytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(bytes, &z.Content)
```

Comment:
Small note: at some point we should try to use the _same line error checks_ just like you do in:

https://github.com/juruen/rmapi/pull/66/files#diff-df27aa6d814a0b1a8eeeae20f1f95890R37-R59

everywhere.

I'm the first one in not being consistent with this 😄 

No need to start doing it now though, it's just a note.

Reply:
The link you provide does not anchor anywhere on the page so I am not sure to understand what you are refering to.
Can you maybe copy the line of code directly in the comment? Thanks.

Reply:
I was just referring to be consistent with how we do error checks when the called function only returns an error:

```go
if err := foo(); err != nil {
   ...
}
```

vs.

```go
err := foo()
if err != nil {
   ...
}
```

Reply:
I agree, I am in favour of merging into one line in the if statement when a called function only returns an error.
I will adjust this one to reflect this.